controllers/cluster_controller_phases.go

```diff
@@ -172,52 +176,56 @@ func (r *ClusterReconciler) reconcileInfrastructure(ctx context.Context, cluster
 
 	if !ready {
 		logger.V(3).Info("Infrastructure provider is not ready yet")
-		return nil
+		return ctrl.Result{}, nil
 	}
 
 	// Get and parse Spec.ControlPlaneEndpoint field from the infrastructure provider.
 	if cluster.Spec.ControlPlaneEndpoint.IsZero() {
 		if err := util.UnstructuredUnmarshalField(infraConfig, &cluster.Spec.ControlPlaneEndpoint, "spec", "controlPlaneEndpoint"); err != nil {
-			return errors.Wrapf(err, "failed to retrieve Spec.ControlPlaneEndpoint from infrastructure provider for Cluster %q in namespace %q",
+			return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve Spec.ControlPlaneEndpoint from infrastructure provider for Cluster %q in namespace %q",
 				cluster.Name, cluster.Namespace)
 		}
 	}
 
 	// Get and parse Status.FailureDomains from the infrastructure provider.
 	if err := util.UnstructuredUnmarshalField(infraConfig, &cluster.Status.FailureDomains, "status", "failureDomains"); err != nil && err != util.ErrUnstructuredFieldNotFound {
-		return errors.Wrapf(err, "failed to retrieve Status.FailureDomains from infrastructure provider for Cluster %q in namespace %q",
+		return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve Status.FailureDomains from infrastructure provider for Cluster %q in namespace %q",
 			cluster.Name, cluster.Namespace)
 	}
 
-	return nil
+	return ctrl.Result{}, nil
 }
 
 // reconcileControlPlane reconciles the Spec.ControlPlaneRef object on a Cluster.
-func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, cluster *clusterv1.Cluster) error {
+func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
 	if cluster.Spec.ControlPlaneRef == nil {
-		return nil
+		return ctrl.Result{}, nil
 	}
 
 	// Call generic external reconciler.
 	controlPlaneReconcileResult, err := r.reconcileExternal(ctx, cluster, cluster.Spec.ControlPlaneRef)
 	if err != nil {
-		return err
+		return ctrl.Result{}, err
+	}
+	// Return early if we need to requeue.
+	if controlPlaneReconcileResult.RequeueAfter > 0 {
+		return ctrl.Result{RequeueAfter: controlPlaneReconcileResult.RequeueAfter}, nil
```

Comment:
Any reason not to just return `controlPlaneReconcileResult` here?
```suggestion
	if !controlPlaneReconcileResult.IsZero {
		return controlPlaneReconcileResult, nil
```

Reply:
They're different, see https://github.com/kubernetes-sigs/cluster-api/pull/3445/files#diff-e167d9e41015e0cc9cb91edb76fbf6a5R28-R34, we'll fix this in v1alpha4 because it'd be a breaking change to rename the external output struct field.